src/VisualStudio/Core/Def/Implementation/Utilities/ServiceInitializer.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Utilities
+{
+    internal class ServiceInitializer<TInterface, TService> where TInterface : class
+    {
+        private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+        private readonly Shell.IAsyncServiceProvider _serviceProvider;
+        private readonly bool _uiThreadRequired;
+
+        private Func<TInterface, Task> _initializer;
+        private TInterface _service;
+
+        public ServiceInitializer(Shell.IAsyncServiceProvider serviceProvider, Func<TInterface, Task> initializer, bool uiThreadRequired)
+        {
+            _serviceProvider = serviceProvider;
+            _initializer = initializer;
+            _uiThreadRequired = uiThreadRequired;
+        }
+
+        public TInterface GetService(CancellationToken cancellationToken)
+        {
+            if (_service == null)
+            {
+                ThreadHelper.JoinableTaskFactory.Run(() => GetServiceAsync(cancellationToken));
+            }
+
+            return _service;
+        }
+
+        public async Task<TInterface> GetServiceAsync(CancellationToken cancellationToken)
+        {
+            var continueOnCapturedContext = _uiThreadRequired;
+
+            Func<TInterface, Task> initializer = null;
+            using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(continueOnCapturedContext))
+            {
+                if (_service != null)
+                {
+                    return _service;
+                }
+
+                initializer = _initializer;
+                _initializer = null;
+            }
+
+            if (_uiThreadRequired)
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            }
+
+            var service = (TInterface)await _serviceProvider.GetServiceAsync(typeof(TService)).ConfigureAwait(continueOnCapturedContext);
+
+            if (initializer != null)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // initializer is not cancellable. otherwise, we can get into unknown state
+                await initializer(service).ConfigureAwait(continueOnCapturedContext);
+            }
+
+            using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(continueOnCapturedContext))
+            {
+                _service = service;
+            }
+
```

Comment:
Notwithstanding your many gate entrances and exits, you have race conditions here, and I don't think you're protecting what you mean to either.
Consider the case where two callers concurrently execute this method. One of them will acquire the `_initializer` value while the other will get null back. They'll both acquire the service, then *one* will execute the initializer while the other will *skip* that step and return the result to the caller. Result: one caller will get the service back without having its initialization complete.

Fixing this without risk of deadlock requires careful code authoring, and is exactly what vs-threading's `AsyncLazy<T>` class is designed to do. I strongly encourage you to use that instead of trying to write your own.

Reply:
good point. I looked your AsyncLazy. it doesn't have GetValue, so are you saying, I get GetService as it is, but use AsyncLazy.GetValueAsync for GetServiceAsync?

Reply:
> it doesn't have GetValue

Ha. I anticipated your desire for a sync version of the method and [add one here](https://github.com/Microsoft/vs-threading/pull/317) for 16.0.
But certainly in the meantime this should not block you. You can always call `AsyncLazy<T>.GetValueAsync`, wrapped in a JTF.Run call.

What I'm proposing is that you move all your thread affinitized initialization logic into an `AsyncLazy<T>` value factory, then have simple Initialize/InitializeAsync methods that invoke the async lazy in order for the side effect to happen.